sql-plugin/src/main/scala/org/apache/spark/sql/rapids/mathExpressions.scala

```diff
@@ -419,6 +419,128 @@ case class GpuCot(child: Expression) extends GpuUnaryMathExpression("COT") {
+      withResource(y.equalTo(value)) { yIsVal =>
+        xIsVal.or(yIsVal)
+      }
+    }
+  }
+
+  def both(value: Scalar, x: ColumnVector, y: ColumnVector): ColumnVector = {
+    withResource(x.equalTo(value)) { xIsVal =>
+      withResource(y.equalTo(value)) { yIsVal =>
+        xIsVal.and(yIsVal)
+      }
+    }
+  }
+
+  def eitherNan(x: ColumnVector, 
+                y: ColumnVector): ColumnVector = {
+    withResource(x.isNan) { xIsNan =>
+      withResource(y.isNan) { yIsNan =>
+        xIsNan.or(yIsNan)
+      }
+    }
+  }
+
+  def eitherNull(x: ColumnVector,
+                 y: ColumnVector): ColumnVector = {
+    withResource(x.isNull) { xIsNull =>
+      withResource(y.isNull) { yIsNull =>
+        xIsNull.or(yIsNull)
+      }
+    }
+  }
+
+  def computeHypot(x: ColumnVector, y: ColumnVector): ColumnVector = {
+    withResource(y.div(x)) { yOverX =>
+      withResource(yOverX.mul(yOverX)) { yOverXSquared =>
+        withResource(Scalar.fromDouble(1)) { one =>
+          withResource(one.add(yOverXSquared)) { onePlusTerm =>
+            withResource(onePlusTerm.sqrt) { onePlusTermSqrt =>
+              x.mul(onePlusTermSqrt)
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+case class GpuHypot(left: Expression, right: Expression) extends CudfBinaryMathExpression("HYPOT") {
+
+  override def binaryOp: BinaryOp = BinaryOp.ADD
+
+  // naive implementation of hypot
+  // r = sqrt(lhs^2 + rhs^2) 
+  // This is prone to overflow with regards either square term
+  // However, we can reduce it to
+  // r = sqrt(x^2 + y^2) = x * sqrt(1 + (y/x)^2)
+  // where x = max(abs(lhs), abs(rhs)), y = min(abs(lhs), abs(rhs))
+  // which will only overflow if both terms are near the maximum representation space
+  override def doColumnar(lhs: GpuColumnVector, rhs: GpuColumnVector): ColumnVector = {
+
+    // in spark SQL HYPOT (java.math.Math.hypot), there are a couple of edge cases that produce
+    // specific results. Note that terms are absolute values (always positive)
+    //  - if either term is inf, then the answer is inf
+    //  - if either term is nan, and neither is inf, then the answer is nan
+    //  - if both terms are 0, then the answer is 0
+
+    val litForSpecial = Seq(Scalar.fromDouble(0),
+                            Scalar.fromDouble(Double.PositiveInfinity),
+                            Scalar.fromDouble(Double.NaN))
+    withResource(litForSpecial) { case Seq(zero, inf, nan) =>
+      withResource(GpuHypot.chooseXandY(lhs.getBase, rhs.getBase)) { case Seq(x, y) =>
+        withResource(Scalar.fromNull(x.getType)) { nullS => 
+          withResource(GpuHypot.either(inf, x, y)) { eitherInf =>
+            withResource(GpuHypot.eitherNull(x, y)) { eitherNull =>
+              withResource(GpuHypot.eitherNan(x, y)) { eitherNan =>
+                withResource(GpuHypot.both(zero, x, y)) { bothZero =>
+                  withResource(GpuHypot.computeHypot(x, y)) { hypotComputed =>
```

Comment:
nit: here we could do the same king of thing to reduce total memory usage.

```
withResource(litForSpecial) { case Seq(zero, inf, nan) =>
  withResource(GpuHypot.chooseXandY(lhs.getBase, rhs.getBase)) { case Seq(x, y) =>
    val zeroOrBase = withResource(GpuHypot.both(zero, x, y)) { bothZero =>
      withResource(GpuHypot.computeHypot(x, y)) { hypotComputed =>
        bothZero.ifElse(zero, hypotComputed)
      }
    }
    val nanOrRest = withResource(zeroOrBase) { zeroOrBase =>
      withResource(GpuHypot.eitherNan(x, y)) { eitherNan =>
        eitherNan.ifElse(nan, zeroOrBase)
      }
    }
    val infOrRest = withResource(nanOrRest) { nanOrRest =>
      withResource(GpuHypot.either(inf, x, y)) { eitherInf =>
        eitherInf.ifElse(inf, nanOrRest)
      }
    }
    withResource(infOrRest) { infOrRest =>
      withResource(GpuHypot.eitherNull(x, y)) { eitherNull =>
        withResource(Scalar.fromNull(x.getType)) { nullS => 
          eitherNull.ifElse(nullS, infOrRest)
        }
      }
    }
  }
}
```

Reply:
Cleaned up in latest commit